File: cpp/medium/FindOrder.cpp
#include <queue>
#include <vector>
using namespace std;

vector<int> findOrder(int numCourses, vector<vector<int>> &prerequisites) {
  vector<int> in(numCourses, 0);
  vector<vector<int>> out(numCourses);
  // init
  for (int i = 0; i < prerequisites.size(); ++i) {
    in[prerequisites[i][0]]++;
    out[prerequisites[i][1]].push_back(prerequisites[i][0]);
  }

  queue<int> q;
  for (int i = 0; i < numCourses; ++i) {
    if (in[i] == 0)
      q.push(i);
  }
  int count = q.size();
  vector<int> rs;
  while (!q.empty()) {
    int temp = q.front();
    q.pop();
    rs.push_back(temp);
    for (auto i : out[temp]) {
      in[i]--;
      if (in[i] == 0)
        q.push(i);
    }
  }
  if (rs.size() == numCourses)
    return rs;
  return {};
}

File: go/middle/findOrder.go
package main

func findOrder(numCourses int, prerequisites [][]int) []int {
	child := make([][]int, numCourses)
	indegree := make([]int, numCourses)
	for i := range prerequisites {
		child[prerequisites[i][1]] = append(child[prerequisites[i][1]], prerequisites[i][0])
		indegree[prerequisites[i][0]]++
	}
	queue := make([]int, 0)
	for i := range indegree {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	rs := make([]int, 0)
	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		rs = append(rs, cur)
		for i := range child[cur] {
			indegree[child[cur][i]]--
			if indegree[child[cur][i]] == 0 {
				queue = append(queue, child[cur][i])
			}
		}
	}
	return rs
}
